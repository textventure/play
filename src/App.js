import React, { Component } from 'react';
import './App.css';

class App extends Component {
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">textventure-play</h1>
        </header>
      </div>
    );
  }
}

if (process.env.NODE_ENV === 'development') {
  const { hot } = require('react-hot-loader');
  App = hot(module)(App);
}

export default App;
