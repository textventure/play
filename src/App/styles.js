import { branchClass } from '../Branch';
const branchSelector = '.' + branchClass;

/**
 * App styles.
 *
 * @param  {Object} theme
 * @return {Object}
 */
const styles = theme => {
  const { palette, spacing, typography } = theme;
  const { pxToRem } = typography;
  const marginBottom1 = pxToRem(spacing(1));
  const marginBottom2 = pxToRem(spacing(2));

  return {
    '@global': {
      [branchSelector]: {
        '& h1': {
          ...typography.display3,
          marginBottom: marginBottom2,
        },
        '& h2': {
          ...typography.display2,
          marginBottom: marginBottom2,
        },
        '& h3': {
          ...typography.display1,
          marginBottom: marginBottom2,
        },
        '& h4': {
          ...typography.headline,
          marginBottom: marginBottom2,
        },
        '& h5': {
          ...typography.title,
          marginBottom: marginBottom2,
        },
        '& h6': {
          ...typography.subheading,
          marginBottom: marginBottom2,
        },
        '& p': {
          ...typography.body1,
          marginBottom: marginBottom1,
        },
      },

      code: {
        fontFamily: 'Consolas, "Liberation Mono", Menlo, Courier, monospace',
        border: `1px solid ${palette.grey[400]}`,
        borderRadius: 3,
        padding: '3px 6px',
      },
      em: {
        fontStyle: 'italic',
      },
      main: {
        display: 'block', // fix for IE 9-11
      },
      strong: {
        fontWeight: typography.fontWeightMedium,
      },
    },

    grid: {
      maxWidth: 720,
    },

    progress: {
      marginTop: spacing(4), // 32px
    },
  };
};

export default styles;
