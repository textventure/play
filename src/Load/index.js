import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Card from '../Card';
import { getStory } from '../helpers/api';

const initialState = {
  error: '',
  value: '',
};
const placeholder = process.env.PUBLIC_URL + '/demo.yaml';

export default class Load extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ...initialState,
      branches: null,
      config: null,
      message: '',
    };

    try {
      const url = new URLSearchParams(props.location.search).get('url');
      if (url) {
        this.state.value = url;
      }
    } catch (error) {}
  }

  /**
   * Handles snackbar close.
   */
  handleClose = () => {
    this.setState({
      message: '',
    });
  };

  /**
   * Handles input change.
   *
   * @param {SyntheticEvent} event
   */
  handleChange = event => {
    const { value } = event.target;
    const newState = {
      ...initialState,
      value,
    };
    try {
      new URL(value);
    } catch (error) {
      newState.error = 'Invalid URL';
    }
    this.setState(newState);
  };

  /**
   * Handles form submit.
   *
   * @param {SyntheticEvent} event
   */
  handleSubmit = event => {
    event.preventDefault();

    getStory(this.state.value)
      .then(story => {
        const { _config: config, ...branches } = story;

        if (config && branches) {
          this.setState({
            branches,
            config,
          });
        }
      })
      .catch(error => {
        this.setState({
          message: error.message,
        });
      });
  };

  render() {
    const { branches, config, value } = this.state;
    if (branches && config) {
      return (
        <Redirect
          to={{
            branches,
            config,
            pathname: '/play',
            search: `url=${encodeURIComponent(value)}`,
          }}
        />
      );
    }

    const { error, message } = this.state;
    return (
      <Card>
        {message && (
          <Snackbar
            autoHideDuration={6000}
            message={message}
            onClose={this.handleClose}
            open={Boolean(message)}
          />
        )}

        <form onSubmit={this.handleSubmit}>
          <Typography gutterBottom variant="headline">
            Load Story
          </Typography>

          <TextField
            autoFocus
            error={Boolean(error)}
            fullWidth
            helperText={error}
            InputLabelProps={{ shrink: true }}
            label="URL"
            margin="normal"
            name="url"
            onChange={this.handleChange}
            placeholder={placeholder}
            required
            spellCheck={false}
            type="url"
            value={value}
          />
          <br />
          <br />

          <Button
            color="primary"
            size="medium"
            type="submit"
            variant="contained"
          >
            Load
          </Button>
        </form>
      </Card>
    );
  }
}
