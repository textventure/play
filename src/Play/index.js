import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';
import LinearProgress from '@material-ui/core/LinearProgress';
import Branch from '../Branch';
import Load from '../Load';
import { getStory } from '../helpers/api';
import { getKey, getValue } from '../helpers/util';

export const defaultConfig = {
  renderer: 'text',
  start: 'start',
};

const styles = theme => ({
  progress: {
    marginTop: theme.spacing.unit * 4, // 32px
  },
});

class Play extends Component {
  static defaultProps = {
    location: {},
  };

  constructor(props) {
    super(props);
    const { branches, config } = props.location;

    this.state = {
      branches,
      config: {
        ...defaultConfig,
        ...config,
      },
      isLoading: true,
    };
    this.state.currentBranchId = this.state.config.start;
  }

  componentDidMount() {
    try {
      const url = new URLSearchParams(window.location.search).get('url');

      getStory(url)
        .then(story => {
          const newState = {
            isLoading: false,
          };

          const { _config: config, ...branches } = story;
          if (branches && config) {
            newState.branches = branches;
            newState.config = {
              ...defaultConfig,
              ...config,
            };
            // set starting branch id
            if (config.start) {
              newState.currentBranchId = config.start;
            }
          }

          this.setState(newState);
        })
        .catch(error => {
          this.setState({
            isLoading: false,
          });
        });
    } catch (error) {}
  }

  /**
   * Updates next story branch id based on choice.
   *
   * @param {String} branchId
   */
  selectChoice = branchId => {
    this.setState({
      currentBranchId: branchId,
    });
  };

  render() {
    const { classes } = this.props;
    const { branches, config, currentBranchId, isLoading } = this.state;

    if (isLoading) {
      return (
        <div className={classes.progress}>
          <LinearProgress />
        </div>
      );
    }

    if (!branches || !config) {
      return <Load />;
    }

    const currentBranch = branches[currentBranchId];

    return (
      <main>
        {currentBranch && (
          <Branch
            choices={getValue(currentBranch)}
            config={config}
            selectChoice={this.selectChoice}
            text={getKey(currentBranch)}
          />
        )}
      </main>
    );
  }
}

export default withStyles(styles)(Play);
