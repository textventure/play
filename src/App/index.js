import React, { Component } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import Grid from '@material-ui/core/Grid';

import Load from '../Load';
import Play from '../Play';

const styles = {
  gridItem: {
    maxWidth: 600,
  },
};

class App extends Component {
  render() {
    const { classes } = this.props;
    return (
      <Router>
        <CssBaseline>
          <Grid container justify="center">
            <Grid item xs={12} className={classes.gridItem}>
              <Switch>
                <Route exact path="/" component={Load} />
                <Route path="/load" component={Load} />
                <Route path="/play" component={Play} />
              </Switch>
            </Grid>
          </Grid>
        </CssBaseline>
      </Router>
    );
  }
}

App = withStyles(styles)(App);

/* istanbul ignore if */
if (process.env.NODE_ENV === 'development') {
  const { hot } = require('react-hot-loader');
  App = hot(module)(App);
}

export default App;
