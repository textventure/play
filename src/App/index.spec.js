import React from 'react';
import { shallow } from 'enzyme';
import App from '.';
import Load from '../Load';
import Play from '../Play';

let wrapper;

beforeAll(() => {
  wrapper = shallow(<App />).dive();
});

it('renders <Load> when path="/"', () => {
  expect(wrapper.find({ path: '/' }).prop('component')).toBe(Load);
});

it('renders <Load> when path="/load"', () => {
  expect(wrapper.find({ path: '/load' }).prop('component')).toBe(Load);
});

it('renders <Play> when path="/play"', () => {
  expect(wrapper.find({ path: '/play' }).prop('component')).toBe(Play);
});

it('renders correctly', () => {
  expect(wrapper.getElement()).toMatchSnapshot();
});
