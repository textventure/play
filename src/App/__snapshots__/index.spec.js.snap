// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<BrowserRouter>
  <GridWrapper
    container={true}
    justify="center"
  >
    <WithStyles(CssBaseline) />
    <GridWrapper
      className="App-gridItem-1"
      item={true}
      xs={12}
    >
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/load"
        />
        <Route
          component={[Function]}
          path="/play"
        />
      </Switch>
    </GridWrapper>
  </GridWrapper>
</BrowserRouter>
`;
